tests/interpreters/test_java.py

```diff
@@ -288,10 +288,10 @@ def test_depth_threshold_with_bin_expr():
     expected_code = """
 public class Model {
     public static double score(double[] input) {
-        return (1.0) + ((1.0) + (subroutine0(input)));
+        return 1.0 + (1.0 + subroutine0(input));
     }
     public static double subroutine0(double[] input) {
-        return (1.0) + ((1.0) + (1.0));
+        return 1.0 + (1.0 + 1.0);
```

Comment:
Sorry, I can't get it. Why do we have braces here (and in some other places like `var0 = (1.0 + 1.0);`)? Is it due to GE here?
```
wrap |= right_precedence is not None and right_precedence >= expr.precedence
```
https://github.com/BayesWitnesses/m2cgen/pull/511/files#diff-5a4f9c32da5612f71e48dad842ece55b645da127a61b6df8dee1c03ad3db76f1R208

Reply:
That's an excellent observation!
> Is it due to GE here?

You're absolutely right. Let me share some additional context.
The expression of kind:
```
1.0 + (1.0 + 1.0)
```
is produced as a result of the AST that looks like following:
```
ast.BinNumExpr(
    left=ast.NumVal(1.0),
    right=ast.BinNumExpr(
        left=ast.NumVal(1.0),
        right=ast.NumVal(1.0),
        op=ast.BinNumOpType.ADD
    ),
    op=ast.BinNumOpType.ADD
)
```
If the expression looked like this:
```
ast.BinNumExpr(
    left=ast.BinNumExpr(
        left=ast.NumVal(1.0),
        right=ast.NumVal(1.0),
        op=ast.BinNumOpType.ADD
    ),
    right=ast.NumVal(1.0),
    op=ast.BinNumOpType.ADD
)
```
You would've seen what you expected:
```
1.0 + 1.0 + 1.0
```
It may seem like this kind of difference is weird and it is for associative operations like addition or multiplication, but consider a slightly modified version of the first tree:
```
```
ast.BinNumExpr(
    left=ast.NumVal(1.0),
    right=ast.BinNumExpr(
        left=ast.NumVal(1.0),
        right=ast.NumVal(1.0),
        op=ast.BinNumOpType.SUB
    ),
    op=ast.BinNumOpType.SUB
)
```
This AST produces the following:
```
1.0 - (1.0 - 1.0)
```
Which makes a lot more sense now - omitting parenthesis would've been completely wrong in this case.

We can still get rid of the remaining parenthesis if we optimize our trees to look like 2nd tree rather than the 1st one for associative operation. Or we can introduce some sort of `is_associative` flag to our expressions which I found to be rather redundant. 

Let me know what you think.

Reply:
Another alternative is to instead of having just `precedence` for each expression we can instead track `left_precedence` and `right_precedence` for each expression individually and get rid of `>=` comparison in the code snippet that you pointed out. 
For associative operators the `left_precedence` and `right_precedence` would have the same value, while for non-associative ones the `right_precedence` will be lower than the `left_precedence`.

Reply:
Now when I thought about it more I really like the latter approach because it also generalizes well to right associative operators (like eg. `**` in Haskell). Let me know if you have any concerns and if not I'll update the implementation.

Reply:
Ah, turns out my idea doesn't really cover scenarios like
```
1.0 - (1.0 + 1.0)
```
Please ignore the above comment.

Reply:
Thank you so much for the detailed explanation with excellent examples! The situation is much clearer now for me. I totally support the idea of introducing some kind of associative rule tracking. But I'd like to propose making further updates in a follow-up PR. This PR is already quite big, self-contained and has already eliminate the great amount of excess braces.

Reply:
Thanks! Merging this then.